rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,129 @@
+
+If the first bytes sent by the server don't match `teleport-resume-v0\r\n` then the client must assume that the server is a regular SSH server, and thus will abort the resumption protocol version exchange, and begin a normal SSH client connection; as `SSH-2.0-` was already sent in the connection, eight bytes must be skipped, assuming that the SSH client will send the same. Such an assumption is valid, as the client side of an SSH connection must send `SSH-2.0-<softwareversion>` at the very beginning of the connection.
+
+If the first bytes sent by the client don't match `SSH-2.0-\x00teleport-resume-v0` then the server must assume that the client is a regular SSH client, and thus will abort the resumption protocol version exchange, and begin a normal SSH server connection. No special handling of the connection is required here, as the server side of the exchange is allowed to send CRLF-terminated lines of UTF-8 text that don't begin with `SSH-` and that don't contribute to the SSH connection state.
+
+### Handshake (new connection)
+
+The clients sends `\x00` (likely together with the previous data), and the server generates a _resumption token_ out of 16 random bytes, but with the highest bit of the first byte set. The server sends this resumption token, then its length-prefixed host ID (which is almost always a UUID in string form, but can also be an EC2 instance ID of varying length). The host ID of the server should be used in later reconnection attempts to ensure that the correct server is receiving the new connection, since the first connection might've been to a hostname or public address and thus might eventually resolve to a different server.
+
+### Handshake (reconnection)
+
+The client sends the previously-received _resumption token_, to which the server replies with a `\x01` to confirm that the token is for a connection that can be resumed or `\x00` if the token is unknown (or for a connection that's expired) followed by a connection close. In the latter case, the client should stop further reconnection attempts and treat the connection as closed by the remote side. The extra byte to confirm or deny the validity of the token is necessary to signal to the client that the connection should be treated as invalid - closing the connection will just trigger a new reconnection attempt.
+
+After the acknowledgement, both sides will send a count of how many bytes they have received in the connection.
+
```

Comment:
Have you given this any further thought on how we avoid potential permission bypass issues?  Is there a way we can detect that an administrator closed the connection and thus it should not be allowed to resume?  Or can session locks be validated at this point?

Reply:
We don't have per-session locks but any lock affecting a user, its roles, its active access requests, its login name, its MFA device or its device trust device will kill the connection; depending on the `locking_mode` (`strict` or `best_effort`) connections will also get closed if the node loses its lock watcher for too long.

Reply:
Can you add this protection to the RFD?  I understand it should just work, but I think it's worth highlighting that we are relying on that behavior.

Reply:
[1c91607](https://github.com/gravitational/teleport/pull/33439/commits/1c916077c1e8ead324be31df32caf75c08d0c5ec)